File: rspace++/src/test/scala/MemConcTest.scala
import org.scalatest.funsuite.AnyFunSuite
import rspacePlusPlus.{RSpacePlusPlus, Setup}
import firefly.rtypes.{Commit, Entry, Retrieve}
import com.sun.jna._
import java.io.File

class MemConcTest extends AnyFunSuite {
  val _ = System.setProperty("jna.library.path", "./target/release/")
  val lib =
    Native
      .load("rspace_plus_plus", classOf[RSpacePlusPlus[Array]])
      .asInstanceOf[RSpacePlusPlus[Array]]

  def getCityField(entry: Entry): String =
    entry.address.get.city

  def getLastNameField(entry: Entry): String =
    entry.name.get.last

  def getStateField(entry: Entry): String =
    entry.address.get.state

  val spacePtr = lib.space_new();
  val setup    = Setup.apply();

  // In-Memory Concurrent
  test("MemConcProduceMatch") {
    // Consume
    val commit =
      Commit(Seq("friends"), Seq(setup.cityMatchCase), "I am the continuation, for now...");
    val commit_buf = commit.toByteArray;
    val cres       = lib.space_put_once_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    // Produce
    val retrieve     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve_buf = retrieve.toByteArray;
    val pres =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve_buf, retrieve_buf.length);

    assert(cres == null)
    assert(!pres.isEmpty())
    assert(lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcProduceNoMatch") {
    // Consume
    val commit =
      Commit(Seq("friends"), Seq(setup.cityMatchCase), "I am the continuation, for now...");
    val commit_buf = commit.toByteArray;
    val cres       = lib.space_put_once_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    // Produce
    val retrieve     = Retrieve("friends", Some(setup.carol), getCityField(setup.carol));
    val retrieve_buf = retrieve.toByteArray;
    val pres =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve_buf, retrieve_buf.length);

    assert(cres == null)
    assert(pres == null)
    assert(!lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcConsumeMatch") {
    // Produce
    val retrieve     = Retrieve("friends", Some(setup.bob), getLastNameField(setup.bob));
    val retrieve_buf = retrieve.toByteArray;
    val pres =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve_buf, retrieve_buf.length);

    // Consume
    val commit =
      Commit(Seq("friends"), Seq(setup.nameMatchCase), "I am the continuation, for now...");
    val commit_buf = commit.toByteArray;
    val cres       = lib.space_put_once_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    assert(pres == null)
    assert(!cres.isEmpty)
    assert(lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcMultipleChannelsConsumeMatch") {
    // Produce
    val retrieve1     = Retrieve("colleagues", Some(setup.dan), getStateField(setup.dan));
    val retrieve1_buf = retrieve1.toByteArray;
    val pres1 =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve1_buf, retrieve1_buf.length);

    // Produce
    val retrieve2     = Retrieve("friends", Some(setup.erin), getStateField(setup.erin));
    val retrieve2_buf = retrieve2.toByteArray;
    val pres2 =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve2_buf, retrieve2_buf.length);

    // Consume
    val commit =
      Commit(
        Seq("friends", "colleagues"),
        Seq(setup.stateMatchCase, setup.stateMatchCase),
        "I am the continuation, for now..."
      );
    val commit_buf = commit.toByteArray;
    val cres       = lib.space_put_once_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    assert(pres1 == null)
    assert(pres2 == null)
    assert(!cres.isEmpty)
    assert(cres.length == 2)
    assert(lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcConsumePersist") {
    // Consume
    val commit =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit_buf = commit.toByteArray;
    val cres =
      lib.space_put_always_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    assert(cres == null)
    assert(!lib.is_empty(spacePtr));

    // Produce
    val retrieve1     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve1_buf = retrieve1.toByteArray;
    val pres =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve1_buf, retrieve1_buf.length);

    assert(!pres.isEmpty())
    assert(!lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcConsumePersistExistingMatches") {
    // Produce
    val retrieve1     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve1_buf = retrieve1.toByteArray;
    val pres1 =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve1_buf, retrieve1_buf.length);

    assert(pres1 == null)

    // Produce
    val retrieve2     = Retrieve("friends", Some(setup.bob), getCityField(setup.alice));
    val retrieve2_buf = retrieve2.toByteArray;
    val pres2 =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve2_buf, retrieve2_buf.length);

    assert(pres2 == null)

    // Consume
    val commit1 =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit1_buf = commit1.toByteArray;
    val cres1 =
      lib.space_put_always_non_durable_concurrent(spacePtr, commit1_buf, commit1_buf.length);

    assert(cres1.length == 1)
    assert(!lib.is_empty(spacePtr));

    val commit2 =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit2_buf = commit2.toByteArray;
    val cres2 =
      lib.space_put_always_non_durable_concurrent(spacePtr, commit2_buf, commit2_buf.length);

    assert(cres2.length == 1)
    assert(lib.is_empty(spacePtr));

    val commit3 =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit3_buf = commit3.toByteArray;
    val cres3 =
      lib.space_put_always_non_durable_concurrent(spacePtr, commit3_buf, commit3_buf.length);

    assert(cres3 == null)
    assert(!lib.is_empty(spacePtr));

    // Produce
    val retrieve3     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve3_buf = retrieve3.toByteArray;
    val pres3 =
      lib.space_get_once_non_durable_concurrent(spacePtr, retrieve3_buf, retrieve3_buf.length);

    assert(!pres3.isEmpty())
    assert(!lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcProducePersist") {
    // Produce
    val retrieve     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve_buf = retrieve.toByteArray;
    val pres =
      lib.space_get_always_non_durable_concurrent(spacePtr, retrieve_buf, retrieve_buf.length);

    assert(pres == null)
    assert(!lib.is_empty(spacePtr));

    // Consume
    val commit =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit_buf = commit.toByteArray;
    val cres       = lib.space_put_once_non_durable_concurrent(spacePtr, commit_buf, commit_buf.length);

    assert(!cres.isEmpty)
    assert(cres.length == 1)
    assert(!lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }

  test("MemConcProducePersistExistingMatches") {
    // Consume
    val commit1 =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit1_buf = commit1.toByteArray;
    val cres1 =
      lib.space_put_once_non_durable_concurrent(spacePtr, commit1_buf, commit1_buf.length);

    assert(cres1 == null)
    assert(!lib.is_empty(spacePtr));

    // Produce
    val retrieve1     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve1_buf = retrieve1.toByteArray;
    val pres1 =
      lib.space_get_always_non_durable_concurrent(spacePtr, retrieve1_buf, retrieve1_buf.length);

    assert(!pres1.isEmpty())
    assert(lib.is_empty(spacePtr));

    // Produce
    val retrieve2     = Retrieve("friends", Some(setup.alice), getCityField(setup.alice));
    val retrieve2_buf = retrieve2.toByteArray;
    val pres2 =
      lib.space_get_always_non_durable_concurrent(spacePtr, retrieve2_buf, retrieve2_buf.length);

    // Consume
    val commit2 =
      Commit(
        Seq("friends"),
        Seq(setup.cityMatchCase),
        "I am the continuation, for now..."
      );
    val commit2_buf = commit2.toByteArray;
    val cres2 =
      lib.space_put_once_non_durable_concurrent(spacePtr, commit2_buf, commit2_buf.length);

    assert(pres2 == null)
    assert(!cres2.isEmpty)
    assert(!lib.is_empty(spacePtr));
    lib.space_clear(spacePtr);
  }
}

File: rspace++/tests/memconc_test.rs
#[cfg(test)]
mod tests {
    use rspace_plus_plus::setup::Setup;

    #[test]
    fn memconc_test_produce_match() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let commit = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, false);

        let retrieve = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice),
        );
        let pres = memconc.produce(retrieve, false);

        assert!(cres.is_none());
        assert!(pres.is_some());
        assert!(memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_produce_no_match() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let commit = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, false);

        let retrieve = Setup::create_retrieve(
            String::from("friends"),
            setup.carol.clone(),
            Setup::get_city_field(setup.carol),
        );
        let pres = memconc.produce(retrieve, false);

        assert!(cres.is_none());
        assert!(pres.is_none());
        assert!(!memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_consume_match() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let retrieve = Setup::create_retrieve(
            String::from("friends"),
            setup.bob.clone(),
            Setup::get_last_name_field(setup.bob),
        );
        let pres = memconc.produce(retrieve, false);

        let commit = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.name_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, false);

        assert!(pres.is_none());
        assert!(cres.is_some());
        assert!(memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_multiple_channels_consume_match() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let retrieve1 = Setup::create_retrieve(
            String::from("colleagues"),
            setup.dan.clone(),
            Setup::get_state_field(setup.dan),
        );
        let pres1 = memconc.produce(retrieve1, false);

        let retrieve2 = Setup::create_retrieve(
            String::from("friends"),
            setup.erin.clone(),
            Setup::get_state_field(setup.erin),
        );
        let pres2 = memconc.produce(retrieve2, false);

        let commit = Setup::create_commit(
            vec![String::from("friends"), String::from("colleagues")],
            vec![setup.state_match_case.clone(), setup.state_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, false);

        assert!(pres1.is_none());
        assert!(pres2.is_none());
        assert!(cres.is_some());
        assert_eq!(cres.unwrap().len(), 2);
        assert!(memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_consume_persist() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let commit = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, true);

        assert!(cres.is_none());
        assert!(!memconc.is_empty());

        let retrieve = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice),
        );
        let pres = memconc.produce(retrieve, false);

        assert!(pres.is_some());
        assert!(!memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_consume_persist_existing_matches() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let retrieve1 = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice.clone()),
        );
        let _pres1 = memconc.produce(retrieve1, false);

        let retrieve2 = Setup::create_retrieve(
            String::from("friends"),
            setup.bob.clone(),
            Setup::get_city_field(setup.bob),
        );
        let _pres2 = memconc.produce(retrieve2, false);

        let commit1 = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case.clone()],
            String::from("I am the continuation, for now..."),
        );
        let cres1 = memconc.consume(commit1, true);

        assert_eq!(cres1.unwrap().len(), 1);
        assert!(!memconc.is_empty());

        let commit2 = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case.clone()],
            String::from("I am the continuation, for now..."),
        );
        let cres2 = memconc.consume(commit2, true);

        assert_eq!(cres2.unwrap().len(), 1);
        assert!(memconc.is_empty());

        let commit3 = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres3 = memconc.consume(commit3, true);

        assert!(cres3.is_none());
        assert!(!memconc.is_empty());

        let retrieve3 = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice),
        );
        let pres3 = memconc.produce(retrieve3, false);

        assert!(pres3.is_some());
        assert!(!memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_produce_persist() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let retrieve = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice),
        );
        let pres = memconc.produce(retrieve, true);

        assert!(pres.is_none());
        assert!(!memconc.is_empty());

        let commit = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let cres = memconc.consume(commit, false);

        assert!(cres.is_some());
        assert_eq!(cres.unwrap().len(), 1);
        assert!(!memconc.is_empty());

        let _ = memconc.clear();
    }

    #[test]
    fn memconc_test_produce_persist_existing_matches() {
        let setup = Setup::new();
        let memconc = setup.memconc;

        let commit1 = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case.clone()],
            String::from("I am the continuation, for now..."),
        );
        let cres1 = memconc.consume(commit1, false);

        assert!(cres1.is_none());
        assert!(!memconc.is_empty());

        let retrieve1 = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice.clone()),
        );
        let pres1 = memconc.produce(retrieve1, true);

        assert!(pres1.is_some());
        assert!((memconc.is_empty()));

        let retrieve2 = Setup::create_retrieve(
            String::from("friends"),
            setup.alice.clone(),
            Setup::get_city_field(setup.alice),
        );
        let pres2 = memconc.produce(retrieve2, true);

        let commit2 = Setup::create_commit(
            vec![String::from("friends")],
            vec![setup.city_match_case],
            String::from("I am the continuation, for now..."),
        );
        let _cres2 = memconc.consume(commit2, false);

        assert!(pres2.is_none());
        assert!(!memconc.is_empty());

        let _ = memconc.clear();
    }
}
